src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -384,6 +338,8 @@ protected void doStart() throws ElasticsearchException {
                 }
 
                 foundPrimary = true;
+                // we rely today on the primary shortcut to not get into infinite loop in chasing primaries
```

Comment:
+1 to having a TransportPrimaryShardOperationAction that will chase the primary (if I get what you mean correctly)